extensions/panache/mongodb-panache/runtime/src/main/java/io/quarkus/mongodb/panache/PanacheMongoEntityBase.java

```diff
@@ -0,0 +1,884 @@
+package io.quarkus.mongodb.panache;
+
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Stream;
+
+import org.bson.Document;
+
+import com.mongodb.client.MongoCollection;
+import com.mongodb.client.MongoDatabase;
+
+import io.quarkus.mongodb.panache.runtime.MongoOperations;
+import io.quarkus.panache.common.Parameters;
+import io.quarkus.panache.common.Sort;
+import io.quarkus.panache.common.impl.GenerateBridge;
+
+/**
+ * <p>
+ * Represents an entity. If your Mongo entities extend this class they gain auto-generated accessors
+ * to all their public fields, as well as a lot of useful
+ * methods. Unless you have a custom ID strategy, you should not extend this class directly but extend
+ * {@link PanacheMongoEntity} instead.
+ * </p>
+ *
+ * @author Loïc Mathieu - loikeseke@gmail.com
+ * @see PanacheMongoEntity
+ */
+public abstract class PanacheMongoEntityBase {
+
+    // Operations
+
+    /**
+     * Insert this entity in the database.
+     * This will set it's ID field if not already set.
+     *
+     * @see #insert(Iterable)
+     * @see #insert(Stream)
+     * @see #insert(Object, Object...)
+     *      TODO delete and keep only persist ?
+     */
+    public void insert() {
```

Comment:
see my previous comment, it's one of the still open discussion point of the PR.
First I have insert/update semtantic as I didn't know how to make a persist() methods that insert if the object is new and update when it is already in the database.
Then I found a way to make a persist() methods that did this.

Now the open question: do I delete all the insert/update methods and keep only persist ?

Reply:
@emmanuelbernard wanted a single method? I would suggest on the contrary:

- persist: inserts a new entity in the DB: it if has no ID, generate it. if it has one it must not exist in the DB otherwise it's an error
- update: updates an entity in the DB: the entity must already exist in the DB, so its ID must be found

Now, this strategy falls flat where you have an existing entity and you want to change its ID, as I've noted in Panache RX. Because we don't store information about the previous ID. We could fix this by injecting a hidden field that gets populated by the ID when we load entities, and when calling `update` we can check if it differs from the current ID. WDYT?

As to whether an upsert method is required, I would personally call it `persistOrUpdate` or `updateOrPersist` rather than `upsert` or anything ambiguous. If we must have one.

Reply:
@emmanuelbernard up

Reply:
I guess I'm in the camp of Stef and have the persist and update semantic he describes. And wait for the upsert operation a bit later but I'm happy to be convinced otherwise. Because JPA's persist is insert only, I'd go for the trio stef suggest instead of making persist an upsert.

Reply:
I don't think I favour a trio. I said I found persist/update more natural, but I agree they can be replaced by a single "upsert" (not with that name).

Reply:
Right so  in decreasing preference:

1. persist and update with split semantic
1. add a third option named `saveOrUpdate` (maybe `merge`) or something

the one thing I don't know is how popular the upsert operation  is for MongoDB users.

Reply:
Personnaly I always used upsert on MongoDB, as it's concistency can be configured to "eventially consistent" (this is the default I think) I found it more convenient.

Reply:
OK. I think that at some point we have to accept that there's an idiomatic way to do things when it comes to SQL people that differs from Mongo users. SQL people will be used to insert/update being separated, especially if coming from ORM (even though the update is implicit), and Mongo users will complain if they use upsert all the time and we don't provide it.

I guess we can accept that we can please everyone and provide:

- `persist` (inserts and new entity, will complain if the entity is already persisted or if the ID already exists in the DB)
- `update` (updates an already-persisted entity to the DB, will complain if the entity is not in the DB)
- `persistOrUpdate` that does an upsert. We can't use `save` next to `persist` because it's just a guess how they differ without reading the docs. `upsert()` is kind of lame because we're using `persist` and not `insert`, and `upsist` is gabbleduck. `saveOrUpdate` would be admitting that we should have called `persist` `save` ;)